### PR TITLE
CIL Examplar Free components

### PR DESCRIPTION
**Pull request by AlbinSou**
I start this PR but it's not finished. Feel free to discuss the things that should be changed. 

For now, here is what I added

- FeCAM (NCM-like but using malahanobis distance (stores per-class covariance matrices) [arxiv](https://arxiv.org/abs/2309.14062)
- FeCAM update utils (current, memory and oracle)
- NCM update utils (current, memory and oracle)
- Cosine Linear first draft

Here is what I want to add:

- Tests for FeCAM
- Tests for all the update utils
- Change Cosine Linear so that it can have new number of new classes (for now it handles only cases with same number of classes in each exp)
- Test for Cosine Linear
- FeatureAdapters [arxiv](https://arxiv.org/abs/2004.00713) (for NCM, for now could not make it work for LinearSVC) + tests
- Resnet18 adapted for Cifar and Tinyimagenet (without maxpool layer) -> @AntonioCarta Do we have this already ?
- Maybe some examples of using all this in continual-learning-baselines
- A lot of regularization based methods use higher lr and different scheduling strategy in the first task and in the remaining ones. I have made a plugin inheriting from LRSchedulerPlugin to handle this kind of things, I could adapt it a bit and propose it (I was thinking of smt general like a plugin that can switch the scheduler based on the task)

### CODE REVIEW

File: avalanche/models/cosine_layer.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+from avalanche.models import DynamicModule
+
+
+"""
+Implementation of Cosine layer taken and modified from https://github.com/G-U-N/PyCIL
+"""
+
+
+class CosineLinear(nn.Module):
```
Comment (AntonioCarta) [inline avalanche/models/cosine_layer.py:15]: doc

File: avalanche/models/cosine_layer.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class SplitCosineLinear(nn.Module):
+    """
+    This class keeps two Cosine Linear layers, without sigma, and handles the sigma parameter
+    that is common for the two of them. One CosineLinear is for the old classes and the other
+    one is for the new classes
+    """
+
+    def __init__(self, in_features, out_features1, out_features2, sigma=True):
```
Comment (AntonioCarta) [inline avalanche/models/cosine_layer.py:50]: doc

File: avalanche/models/cosine_layer.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+        return out
+
+
+class SplitCosineLinear(nn.Module):
+    """
+    This class keeps two Cosine Linear layers, without sigma, and handles the sigma parameter
+    that is common for the two of them. One CosineLinear is for the old classes and the other
+    one is for the new classes
+    """
```
Comment (AntonioCarta) [inline avalanche/models/cosine_layer.py:48]: what is sigma?

File: avalanche/models/cosine_layer.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        if self.sigma is not None:
+            out = self.sigma * out
+
+        return out
+
+
+class CosineIncrementalClassifier(DynamicModule):
+    # WARNING Maybe does not work with initial evaluation
+    def __init__(self, in_features, num_classes):
```
Comment (AntonioCarta) [inline avalanche/models/cosine_layer.py:76]: num_classes should be optional in the incremental version? (1 is a good default)

File: avalanche/models/cosine_layer.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            fc.fc1.weight.data[:prev_out_features1] = self.fc.fc1.weight.data
+            fc.fc1.weight.data[prev_out_features1:] = self.fc.fc2.weight.data
+            fc.sigma.data = self.fc.sigma.data
+        del self.fc
+        self.fc = fc
+
+    def forward(self, x):
+        return self.fc(x)
+
+    def generate_fc(self, in_dim, out_dim):
```
Comment (AntonioCarta) [inline avalanche/models/cosine_layer.py:103]: internal method? use `_generate_fc`. If it's public add doc.

**Review by AntonioCarta — CHANGES_REQUESTED** [review]

**Comment by AntonioCarta** [comment]
Thanks @AlbinSou. Can we split this in multiple PRs? I think FeCAM and CosineLayer are ready once you add tests and doc. We can add the rest later.

I'm not sure about the plugins. The decoupling between the modules and their logic makes it a bit more complex to use them but I don't have a solution for this. If there is no alternative solution, I would at least add a reference to their implementation in the Module's doc.

File: avalanche/models/cosine_layer.py (hunk 4 above)
Comment (AlbinSou) [reply avalanche/models/cosine_layer.py:76]: About that, I am not sure right now whether this behavior can be implemented with the current code state. In this code, whenever new classes arrive, the classifier switchs from using one CosineLinear layer to a SplitCosineLinear one (containing two CosineLinear). So, in the case we initialize with 1 class, this SplitCosineLinear will contain the old layer (with 1 logit) plus a new layer containing the new seen classes (let say 9 new ones for the first task). I think this behavior is a bit weird. But I agree that it would be nice to be consistent with the IncrementalClassifier behavior, I will make some more tests to make sure we can get a version that is both correct and consistent with the current IncrementalClassifier

Comment (AntonioCarta) [reply avalanche/models/cosine_layer.py:76]: If I understand correctly, this means that you should not have a `num_classes` parameter at all. You must always start with 0 units.
